File: code/matlab/chapter06/mcse.m
function MCSE = mcse(draws, L)
% mcse.m
% Computes Monte Carlo standard errors (MCSEs) for posterior means
% obtained from MCMC output. For each column of draws, MCSE(j) is
% sqrt(Omega_j / R), where Omega_j is the long-run variance estimated
% by the spectral variance at zero. Requires specvar0.m.
%
% Inputs:
%   draws : R-by-k matrix of MCMC draws
%   L     : truncation lag for the spectral variance estimator
%
% Output:
%   MCSE  : 1-by-k vector of Monte Carlo standard errors
[R, k] = size(draws);
MCSE = zeros(1, k);
for j = 1:k
    x = draws(:, j);
    Omega = specvar0(x, L) * R; % long-run variance of x^{(r)}
    MCSE(j) = sqrt(Omega / R);
end
end

File: code/python/chapter06/mcse.py
# mcse.py
# Computes Monte Carlo standard errors (MCSEs) for posterior means
# obtained from MCMC output. For each column of draws, MCSE(j) is
# sqrt(Omega_j / R), where Omega_j is the long-run variance estimated
# by the spectral variance at zero. Requires specvar0.py.

import numpy as np
from specvar0 import specvar0


def mcse(draws, L):
    """Monte Carlo standard errors for each column of draws.

    Inputs:
      draws : R-by-k matrix of MCMC draws
      L     : truncation lag for the spectral variance estimator

    Output:
      MCSE  : k-vector of Monte Carlo standard errors
    """
    R, k = draws.shape
    MCSE = np.zeros(k)
    for j in range(k):
        x = draws[:, j]
        Omega = specvar0(x, L)*R  # long-run variance of x^{(r)}
        MCSE[j] = np.sqrt(Omega/R)
    return MCSE
